0.16.0 breaks with `AttributeError: ImageComparison object has no attribute result`

`pytest-mpl 0.15.0` was working for us, but `pytest-mpl 0.16.0` has broken our tests with an error inside `pytest-mpl`.

On Python 3.9 running on my laptop, an example stack trace is:

```
_________________________________________________________ TimeSeriesDataOpsTest.test_plot _________________________________________________________

self = <pytest_mpl.plugin.ImageComparison object at 0x139932340>, item = <TestCaseFunction test_plot>

    @pytest.hookimpl(hookwrapper=True)
    def pytest_runtest_call(self, item):  # noqa
    
        compare = get_compare(item)
    
        if compare is None:
            yield
            return
    
        import matplotlib.pyplot as plt
        try:
            from matplotlib.testing.decorators import remove_ticks_and_titles
        except ImportError:
            from matplotlib.testing.decorators import ImageComparisonTest as MplImageComparisonTest
            remove_ticks_and_titles = MplImageComparisonTest.remove_text
    
        style = compare.kwargs.get('style', 'classic')
        remove_text = compare.kwargs.get('remove_text', False)
        backend = compare.kwargs.get('backend', 'agg')
    
        with plt.style.context(style, after_reset=True), switch_backend(backend):
    
            # Run test and get figure object
            yield
>           fig = self.result
E           AttributeError: 'ImageComparison' object has no attribute 'result'

/usr/local/lib/python3.9/site-packages/pytest_mpl/plugin.py:611: AttributeError
-------------------------------------------------------------- Captured stderr call ---------------------------------------------------------------
```

On Python 3.7 running on GitHub, this points to `FigureCloser` instead, but otherwise the same error:

```
self = <pytest_mpl.plugin.FigureCloser object at 0x7fa83e0c2dd0>
item = <TestCaseFunction test_plot>
    @pytest.hookimpl(hookwrapper=True)
    def pytest_runtest_call(self, item):
        yield
        if get_compare(item) is not None:
>           close_mpl_figure(self.result)
E           AttributeError: 'FigureCloser' object has no attribute 'result'
/opt/hostedtoolcache/Python/3.7.13/x64/lib/python3.7/site-packages/pytest_mpl/plugin.py:740: AttributeError
```

[Failed Github Action](https://github.com/facebookresearch/Kats/runs/7048290085?check_suite_focus=true)

--- a/pytest_mpl/plugin.py
+++ b/pytest_mpl/plugin.py
@@ -622,6 +622,9 @@ class ImageComparison:
             wrap_figure_interceptor(self, item)
             yield
             test_name = generate_test_name(item)
+            if test_name not in self.return_value:
+                # Test function did not complete successfully
+                return
             fig = self.return_value[test_name]
 
             if remove_text:
@@ -749,5 +752,8 @@ class FigureCloser:
         yield
         if get_compare(item) is not None:
             test_name = generate_test_name(item)
+            if test_name not in self.return_value:
+                # Test function did not complete successfully
+                return
             fig = self.return_value[test_name]
             close_mpl_figure(fig)

--- a/pytest_mpl/plugin.py
+++ b/pytest_mpl/plugin.py
@@ -81,14 +81,6 @@ def pathify(path):
     return Path(path + ext)
 
 
-def _pytest_pyfunc_call(obj, pyfuncitem):
-    testfunction = pyfuncitem.obj
-    funcargs = pyfuncitem.funcargs
-    testargs = {arg: funcargs[arg] for arg in pyfuncitem._fixtureinfo.argnames}
-    obj.result = testfunction(**testargs)
-    return True
-
-
 def generate_test_name(item):
     """
     Generate a unique name for the hash for this test.
@@ -98,6 +90,24 @@ def generate_test_name(item):
     else:
         name = f"{item.module.__name__}.{item.name}"
     return name
+
+
+def wrap_figure_interceptor(plugin, item):
+    """
+    Intercept and store figures returned by test functions.
+    """
+    # Only intercept figures on marked figure tests
+    if get_compare(item) is not None:
+
+        # Use the full test name as a key to ensure correct figure is being retrieved
+        test_name = generate_test_name(item)
+
+        def figure_interceptor(store, obj):
+            def wrapper(*args, **kwargs):
+                store.return_value[test_name] = obj(*args, **kwargs)
+            return wrapper
+
+        item.obj = figure_interceptor(plugin, item.obj)
 
 
 def pytest_report_header(config, startdir):
@@ -286,6 +296,7 @@ class ImageComparison:
         self._generated_hash_library = {}
         self._test_results = {}
         self._test_stats = None
+        self.return_value = {}
 
         # https://stackoverflow.com/questions/51737378/how-should-i-log-in-my-pytest-plugin
         # turn debug prints on only if "-vv" or more passed
@@ -608,13 +619,14 @@ class ImageComparison:
         with plt.style.context(style, after_reset=True), switch_backend(backend):
 
             # Run test and get figure object
+            wrap_figure_interceptor(self, item)
             yield
-            fig = self.result
+            test_name = generate_test_name(item)
+            fig = self.return_value[test_name]
 
             if remove_text:
                 remove_ticks_and_titles(fig)
 
-            test_name = generate_test_name(item)
             result_dir = self.make_test_results_dir(item)
 
             summary = {
@@ -678,10 +690,6 @@ class ImageComparison:
             if summary['status'] == 'skipped':
                 pytest.skip(summary['status_msg'])
 
-    @pytest.hookimpl(tryfirst=True)
-    def pytest_pyfunc_call(self, pyfuncitem):
-        return _pytest_pyfunc_call(self, pyfuncitem)
-
     def generate_summary_json(self):
         json_file = self.results_dir / 'results.json'
         with open(json_file, 'w') as f:
@@ -733,13 +741,13 @@ class FigureCloser:
 
     def __init__(self, config):
         self.config = config
+        self.return_value = {}
 
     @pytest.hookimpl(hookwrapper=True)
     def pytest_runtest_call(self, item):
+        wrap_figure_interceptor(self, item)
         yield
         if get_compare(item) is not None:
-            close_mpl_figure(self.result)
-
-    @pytest.hookimpl(tryfirst=True)
-    def pytest_pyfunc_call(self, pyfuncitem):
-        return _pytest_pyfunc_call(self, pyfuncitem)
+            test_name = generate_test_name(item)
+            fig = self.return_value[test_name]
+            close_mpl_figure(fig)

--- a/pytest_mpl/plugin.py
+++ b/pytest_mpl/plugin.py
@@ -89,6 +89,17 @@ def _pytest_pyfunc_call(obj, pyfuncitem):
     return True
 
 
+def generate_test_name(item):
+    """
+    Generate a unique name for the hash for this test.
+    """
+    if item.cls is not None:
+        name = f"{item.module.__name__}.{item.cls.__name__}.{item.name}"
+    else:
+        name = f"{item.module.__name__}.{item.name}"
+    return name
+
+
 def pytest_report_header(config, startdir):
     import matplotlib
     import matplotlib.ft2font
@@ -287,7 +298,7 @@ class ImageComparison:
         Given a pytest item, generate the figure filename.
         """
         if self.config.getini('mpl-use-full-test-name'):
-            filename = self.generate_test_name(item) + '.png'
+            filename = generate_test_name(item) + '.png'
         else:
             compare = get_compare(item)
             # Find test name to use as plot name
@@ -298,21 +309,11 @@ class ImageComparison:
         filename = str(pathify(filename))
         return filename
 
-    def generate_test_name(self, item):
-        """
-        Generate a unique name for the hash for this test.
-        """
-        if item.cls is not None:
-            name = f"{item.module.__name__}.{item.cls.__name__}.{item.name}"
-        else:
-            name = f"{item.module.__name__}.{item.name}"
-        return name
-
     def make_test_results_dir(self, item):
         """
         Generate the directory to put the results in.
         """
-        test_name = pathify(self.generate_test_name(item))
+        test_name = pathify(generate_test_name(item))
         results_dir = self.results_dir / test_name
         results_dir.mkdir(exist_ok=True, parents=True)
         return results_dir
@@ -526,7 +527,7 @@ class ImageComparison:
             pytest.fail(f"Can't find hash library at path {hash_library_filename}")
 
         hash_library = self.load_hash_library(hash_library_filename)
-        hash_name = self.generate_test_name(item)
+        hash_name = generate_test_name(item)
         baseline_hash = hash_library.get(hash_name, None)
         summary['baseline_hash'] = baseline_hash
 
@@ -613,7 +614,7 @@ class ImageComparison:
             if remove_text:
                 remove_ticks_and_titles(fig)
 
-            test_name = self.generate_test_name(item)
+            test_name = generate_test_name(item)
             result_dir = self.make_test_results_dir(item)
 
             summary = {

--- a/setup.cfg
+++ b/setup.cfg
@@ -47,6 +47,8 @@ markers =
     hash: run test during hash comparison only mode.
 filterwarnings =
     error
+    ignore:distutils Version classes are deprecated
+    ignore:the imp module is deprecated in favour of importlib
 
 [flake8]
 max-line-length = 100

--- a/setup.cfg
+++ b/setup.cfg
@@ -42,6 +42,11 @@ test =
 
 [tool:pytest]
 testpaths = "tests"
+markers =
+    image: run test during image comparison only mode.
+    hash: run test during hash comparison only mode.
+filterwarnings =
+    error
 
 [flake8]
 max-line-length = 100

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -1,0 +1,9 @@
+import pytest
+from packaging.version import Version
+
+pytest_plugins = ["pytester"]
+
+if Version(pytest.__version__) < Version("6.2.0"):
+    @pytest.fixture
+    def pytester(testdir):
+        return testdir

--- a/tests/test_pytest_mpl.py
+++ b/tests/test_pytest_mpl.py
@@ -566,3 +566,107 @@ def test_class_fail(code, tmpdir):
     # If we don't use --mpl option, the test should succeed
     code = call_pytest([test_file])
     assert code == 0
+
+
+@pytest.mark.parametrize("runpytest_args", [(), ("--mpl",)])
+def test_user_fail(pytester, runpytest_args):
+    pytester.makepyfile(
+        """
+        import pytest
+        @pytest.mark.mpl_image_compare
+        def test_fail():
+            pytest.fail("Manually failed by user.")
+    """
+    )
+    result = pytester.runpytest(*runpytest_args)
+    result.assert_outcomes(failed=1)
+    result.stdout.fnmatch_lines("FAILED*Manually failed by user.*")
+
+
+@pytest.mark.parametrize("runpytest_args", [(), ("--mpl",)])
+def test_user_skip(pytester, runpytest_args):
+    pytester.makepyfile(
+        """
+        import pytest
+        @pytest.mark.mpl_image_compare
+        def test_skip():
+            pytest.skip("Manually skipped by user.")
+    """
+    )
+    result = pytester.runpytest(*runpytest_args)
+    result.assert_outcomes(skipped=1)
+
+
+@pytest.mark.parametrize("runpytest_args", [(), ("--mpl",)])
+def test_user_importorskip(pytester, runpytest_args):
+    pytester.makepyfile(
+        """
+        import pytest
+        @pytest.mark.mpl_image_compare
+        def test_importorskip():
+            pytest.importorskip("nonexistantmodule")
+    """
+    )
+    result = pytester.runpytest(*runpytest_args)
+    result.assert_outcomes(skipped=1)
+
+
+@pytest.mark.parametrize("runpytest_args", [(), ("--mpl",)])
+def test_user_xfail(pytester, runpytest_args):
+    pytester.makepyfile(
+        """
+        import pytest
+        @pytest.mark.mpl_image_compare
+        def test_xfail():
+            pytest.xfail()
+    """
+    )
+    result = pytester.runpytest(*runpytest_args)
+    result.assert_outcomes(xfailed=1)
+
+
+@pytest.mark.parametrize("runpytest_args", [(), ("--mpl",)])
+def test_user_exit_success(pytester, runpytest_args):
+    pytester.makepyfile(
+        """
+        import pytest
+        @pytest.mark.mpl_image_compare
+        def test_exit_success():
+            pytest.exit("Manually exited by user.", returncode=0)
+    """
+    )
+    result = pytester.runpytest(*runpytest_args)
+    result.assert_outcomes()
+    assert result.ret == 0
+    result.stdout.fnmatch_lines("*Exit*Manually exited by user.*")
+
+
+@pytest.mark.parametrize("runpytest_args", [(), ("--mpl",)])
+def test_user_exit_failure(pytester, runpytest_args):
+    pytester.makepyfile(
+        """
+        import pytest
+        @pytest.mark.mpl_image_compare
+        def test_exit_fail():
+            pytest.exit("Manually exited by user.", returncode=1)
+    """
+    )
+    result = pytester.runpytest(*runpytest_args)
+    result.assert_outcomes()
+    assert result.ret == 1
+    result.stdout.fnmatch_lines("*Exit*Manually exited by user.*")
+
+
+@pytest.mark.parametrize("runpytest_args", [(), ("--mpl",)])
+def test_user_function_raises(pytester, runpytest_args):
+    pytester.makepyfile(
+        """
+        import pytest
+        @pytest.mark.mpl_image_compare
+        def test_raises():
+            raise ValueError("User code raised an exception.")
+    """
+    )
+    result = pytester.runpytest(*runpytest_args)
+    result.assert_outcomes(failed=1)
+    result.stdout.fnmatch_lines("FAILED*ValueError*User code*")

--- a/tests/test_pytest_mpl.py
+++ b/tests/test_pytest_mpl.py
@@ -3,6 +3,7 @@ import sys
 import json
 import subprocess
 from pathlib import Path
+from unittest import TestCase
 
 import matplotlib
 import matplotlib.ft2font
@@ -247,6 +248,23 @@ class TestClassWithSetup:
     # Regression test for a bug that occurred when using setup_method
 
     def setup_method(self, method):
+        self.x = [1, 2, 3]
+
+    @pytest.mark.mpl_image_compare(baseline_dir=baseline_dir_local,
+                                   filename='test_succeeds.png',
+                                   tolerance=DEFAULT_TOLERANCE)
+    def test_succeeds(self):
+        fig = plt.figure()
+        ax = fig.add_subplot(1, 1, 1)
+        ax.plot(self.x)
+        return fig
+
+
+class TestClassWithTestCase(TestCase):
+
+    # Regression test for a bug that occurred when using unittest.TestCase
+
+    def setUp(self):
         self.x = [1, 2, 3]
 
     @pytest.mark.mpl_image_compare(baseline_dir=baseline_dir_local,
@@ -514,8 +532,27 @@ class TestClassWithSetup:
         return fig
 """
 
+TEST_FAILING_UNITTEST_TESTCASE = """
+from unittest import TestCase
+import pytest
+import matplotlib.pyplot as plt
+class TestClassWithTestCase(TestCase):
+    def setUp(self):
+        self.x = [1, 2, 3]
+    @pytest.mark.mpl_image_compare
+    def test_fails(self):
+        fig = plt.figure()
+        ax = fig.add_subplot(1, 1, 1)
+        ax.plot(self.x)
+        return fig
+"""
 
-@pytest.mark.parametrize("code", [TEST_FAILING_CLASS, TEST_FAILING_CLASS_SETUP_METHOD])
+
+@pytest.mark.parametrize("code", [
+    TEST_FAILING_CLASS,
+    TEST_FAILING_CLASS_SETUP_METHOD,
+    TEST_FAILING_UNITTEST_TESTCASE,
+])
 def test_class_fail(code, tmpdir):
 
     test_file = tmpdir.join('test.py').strpath

--- a/tox.ini
+++ b/tox.ini
@@ -51,8 +51,3 @@ description = check code style, e.g. with flake8
 deps = pre-commit
 commands =
     pre-commit run --all-files
-
-[pytest]
-markers =
-    image: run test during image comparison only mode.
-    hash: run test during hash comparison only mode.

